### PR TITLE
#19477: Expose vector mode options for sigmoid op

### LINKED ISSUE #19477: Speed up sigmoid in cases of non 32x32 tiles.
**Describe the bug**
In a customer model, we have a convolution which outputs a tensor of logical shape [1, 1, 2048*128, 2]. Due to convolution/tt-metal op constraints, this shape will actually get padded to [1, 1, 2048 * 128, 32].
Following the convolution, sigmoid is called.
Sigmoid will operate on a tile-by-tile basis, calculating the results even for the padded last dim.
In this case, the duration of the sigmoid will be somewhere around 220k cycles, amounting for more than 25% device perf of the full model.

Now, ideally, we would not be computing sigmoid for full tiles (32x2 is what is needed).
Sigmoid LLK already has a mechanism which somewhat addresses this, and it's called VectorMode.
Using VectorMode::C (which tells SFPU to ignore faces 1 & 4) drops duration to ~120k cycles, almost a 2x speedup :)

To do that, I've hacked the sigmoid_tile HLK call to always use VectorMode::C, like this:
`ALWI void sigmoid_tile(uint32_t idst) { MATH((llk_math_eltwise_unary_sfpu_sigmoid<APPROX>(idst, 2))); }
`
So, as a first step of optimising sigmoid (and other unary ops will benefit from this, when needed), I think exposing VectorMode on unary ops should be a good start.

Ideally, we would want to perform computation on only 32x2 actual data, which LLK doesn't support ATM.

**To Reproduce**
Attaching a unit test:

```
import pytest

import torch

import ttnn

from tests.ttnn.utils_for_testing import assert_with_pcc, assert_equal
from tests.ttnn.unit_tests.operations.eltwise.backward.utility_funcs import data_gen_with_range, compare_pcc
from models.utility_functions import torch_random, skip_for_grayskull, is_wormhole_b0, is_blackhole


def run_unary_test_sharded(device, hw, out_channels, ttnn_function, pcc=0.9999):
    torch.manual_seed(0)

    torch_input_tensor = torch.rand((hw, out_channels), dtype=torch.bfloat16)
    golden_function = ttnn.get_golden_function(ttnn_function)
    torch_output_tensor = golden_function(torch_input_tensor, device=device)

    num_cores = 64
    core_range_set = ttnn.num_cores_to_corerangeset(num_cores, device.compute_with_storage_grid_size())
    shard_shape = [
        hw // num_cores,
        out_channels if out_channels >= 32 else 32
    ]

    memory_config = ttnn.MemoryConfig(
                ttnn.TensorMemoryLayout.HEIGHT_SHARDED,
                ttnn.BufferType.L1,
                ttnn.ShardSpec(
                    core_range_set,
                    shard_shape,
                    ttnn.ShardOrientation.ROW_MAJOR,
                ),
            )

    input_tensor = ttnn.from_torch(torch_input_tensor, layout=ttnn.TILE_LAYOUT, device=device, memory_config=memory_config)
    output_tensor = ttnn_function(input_tensor)
    output_tensor = ttnn.from_device(output_tensor)
    output_tensor = ttnn.to_torch(output_tensor)

    assert_with_pcc(torch_output_tensor, output_tensor, pcc)

@pytest.mark.parametrize("h", [2048])
@pytest.mark.parametrize("w", [128])
@pytest.mark.parametrize("out_channels", [2])
def test_sigmoid_two_out_channels(device, h, w, out_channels):
    torch.manual_seed(0)
    run_unary_test_sharded(device, h * w, out_channels, ttnn.sigmoid, pcc=0.991)

```
To try VectorMode::C, go to tt_metal/include/compute_kernel_api.h, and modify the implementation to be like this;

Image

That should give a speedup, and maintain PCC accuracy.

### PR DESCRIPTION
**Pull request by umadevimcw**
### Ticket
#19477 

### Problem description
To speed up sigmoid op by using different vector modes

### What's changed
Update the sigmoid op call unary ops related files

### Checklist
- [ ] [All post commit](https://github.com/tenstorrent/tt-metal/actions/runs/14326281430/job/40156287697) CI passes
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/runs/14216083738) CI passes (if applicable)
- [ ] [Model perf](https://github.com/tenstorrent/tt-metal/actions/runs/14222721628)
- [ ] [Device perf](https://github.com/tenstorrent/tt-metal/actions/runs/14222715723)
- [ ] [Nightly run](https://github.com/tenstorrent/tt-metal/actions/runs/14219537090)

For the specific exmaple mentioned in the issue (#19477), please find the numbers below

- sigmoid_tile +  RC = 225600 ns
- sigmoid_tile +  C = 118843 ns
- sigmoid_approx_tile + C = 22705 ns
- sigmoid_approx_tile + RC = 37359 ns

### CODE REVIEW

**Review by KalaivaniMCW — APPROVED** [review]

**Review by Aswinmcw — APPROVED** [review]

File: tt_metal/include/compute_kernel_api.h [hunk 1]
```diff
@@ -89,6 +89,10 @@ ALWI void sigmoid_tile_init() {
     MATH((llk_math_eltwise_unary_sfpu_sigmoid_init<fast_and_approx>()));
 }
 
+template <bool fast_and_approx = true>
```
[the remark this answers is not in the document]
Comment (pavlepopovic) [reply tt_metal/include/compute_kernel_api.h:92]: sigmoid_tile_init() (the function above this), has already <fast_and_approx> template, why add new one?
In the impl below llk_math_eltwise_unary_sfpu_sigmoid/init should, based on the template, pick the approx mode version or not?
can we just use sigmoid_tile_init and expand it, in addition to having approx mode (true or false), to have vector mode (template or parameter)
then we don't need this API right?

[the remark this answers is not in the document]
Comment (pavlepopovic) [reply tt_metal/include/compute_kernel_api.h:92]: Think that then we don't need llk include changes above I think

File: ttnn/cpp/ttnn/operations/eltwise/unary/common/unary_op_utils.cpp [hunk 2]
```diff
@@ -170,6 +170,20 @@ std::pair<std::string, std::string> get_op_init_and_func_parameterized(
...
+            TT_FATAL(param0 == 2 || param0 == 4, "Sigmoid accepts only vector mode C and RC");
+            auto param1 = (uint32_t)params[1];
+            if (param1) {
+                op_init_and_name = {
+                    fmt::format("sigmoid_approx_tile_init<{}u>();", param1),
+                    fmt::format("sigmoid_approx_tile<{1}, {2}u>({0});", idst, (int32_t)param0, param1)};
+            } else {
+                op_init_and_name = {
+                    fmt::format("sigmoid_tile_init<{}u>();", param1),
+                    fmt::format("sigmoid_tile<{1}u>({0});", idst, param1)};
```
Comment (pavlepopovic) [inline ttnn/cpp/ttnn/operations/eltwise/unary/common/unary_op_utils.cpp:183]: this here is calling sigmoid_tile_init<approx_mode> already

File: tests/ttnn/unit_tests/operations/eltwise/test_unary_ops_ttnn.py [hunk 3]
```diff
@@ -852,7 +852,9 @@ def test_unary_sigmoid_ttnn(input_shapes, device, approx_mode):
     _, output_tensor = data_gen_with_range(input_shapes, -1, 1, device)
 
     cq_id = 0
-    ttnn.sigmoid(input_tensor, fast_and_approximate_mode=approx_mode, output_tensor=output_tensor, queue_id=cq_id)
+    ttnn.sigmoid(
+        input_tensor, vector_mode=4, fast_and_approximate_mode=approx_mode, output_tensor=output_tensor, queue_id=cq_id
```
Comment (pavlepopovic) [inline tests/ttnn/unit_tests/operations/eltwise/test_unary_ops_ttnn.py:856]: can we add a test here targeting vector mode C as well?
In addition to that, can we add test mentioned in issue https://github.com/tenstorrent/tt-metal/issues/19477 here as well? That one should verify that vector mode C is actually doing what we expect?

**Review by VirdhatchaniKN — APPROVED** [review]

File: tests/ttnn/unit_tests/operations/eltwise/test_unary_ops_ttnn.py (hunk 3 above)
Comment (umadevimcw) [reply tests/ttnn/unit_tests/operations/eltwise/test_unary_ops_ttnn.py:856]: @pavlepopovic  I did add the test mentioned in the issue #19477, but i was encountering issue in WHB0 N300 CI test. Number of cores hard coded as 64 in test which is not working for N300 

image

File: tt_metal/include/compute_kernel_api.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (umadevimcw) [reply tt_metal/include/compute_kernel_api.h:92]: @pavlepopovic LLK include changes are required since the approximate implementation in separate file.

**Review by mouliraj-mcw — APPROVED** [review]

File: tt_metal/include/compute_kernel_api.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (umadevimcw) [reply tt_metal/include/compute_kernel_api.h]: Updated the code

File: tests/ttnn/unit_tests/operations/eltwise/test_unary_ops_ttnn.py (hunk 3 above)
Comment (umadevimcw) [reply tests/ttnn/unit_tests/operations/eltwise/test_unary_ops_ttnn.py:856]: Added the test file, but have to figure out to fix the above issue

File: tests/ttnn/unit_tests/operations/eltwise/test_unary.py [hunk 4]
```diff
@@ -37,7 +37,7 @@ def run_unary_with_approx_mode_test(device, h, w, ttnn_function, approx_mode, pc
     torch_output_tensor = golden_function(torch_input_tensor, device=device)
 
     input_tensor = ttnn.from_torch(torch_input_tensor, layout=ttnn.TILE_LAYOUT, device=device)
-    output_tensor = ttnn_function(input_tensor, fast_and_approximate_mode=approx_mode)
+    output_tensor = ttnn_function(input_tensor, vector_mode=4, fast_and_approximate_mode=approx_mode)
```
[the remark this answers is not in the document]
Comment (umadevimcw) [reply tests/ttnn/unit_tests/operations/eltwise/test_unary.py:40]: Updated the test file

File: ttnn/cpp/ttnn/operations/eltwise/unary/unary.hpp [hunk 5]
```diff
@@ -43,6 +43,17 @@ struct ExecuteUnaryWithFastAndApproximateMode {
         const std::optional<Tensor>& optional_output_tensor = std::nullopt);
 };
 
+template <UnaryOpType unary_op_type>
+struct ExecuteUnaryWithVectorAndFastAndApproximateMode {
+    static Tensor invoke(
+        QueueId queue_id,
+        const Tensor& input_tensor,
+        const int mode = 4,
```
[the remark this answers is not in the document]
Comment (umadevimcw) [reply ttnn/cpp/ttnn/operations/eltwise/unary/unary.hpp]: updated

File: tt_metal/include/compute_kernel_api.h [hunk 6]
```diff
@@ -108,6 +112,12 @@ ALWI void sigmoid_tile(uint32_t idst) {
     MATH((llk_math_eltwise_unary_sfpu_sigmoid<fast_and_approx>(idst)));
 }
 
+// clang-format on
+// defaults to RC mode
+template <int vec_mode = 4, bool fast_and_approx = false>
```
[the remark this answers is not in the document]
Comment (umadevimcw) [reply tt_metal/include/compute_kernel_api.h]: updated

File: ttnn/cpp/ttnn/operations/eltwise/unary/common/unary_op_utils.cpp (hunk 2 above)
Comment (umadevimcw) [reply ttnn/cpp/ttnn/operations/eltwise/unary/common/unary_op_utils.cpp]: updated the code

**Review by pavlepopovic — APPROVED** [review]

File: tests/ttnn/unit_tests/operations/eltwise/test_unary_ops_ttnn.py (hunk 3 above)
Comment (umadevimcw) [reply tests/ttnn/unit_tests/operations/eltwise/test_unary_ops_ttnn.py:856]: added thanks.

**Review by mcw-anasuya — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/eltwise/unary_backward/unary_backward.cpp [hunk 7]
```diff
@@ -510,7 +510,9 @@ std::vector<Tensor> ExecuteUnaryBackwardTan::invoke(
 std::vector<Tensor> ExecuteUnaryBackwardSigmoid::invoke(
     const Tensor& grad, const Tensor& input, const std::optional<MemoryConfig>& output_mem_config) {
     std::vector<Tensor> grad_tensor;
-    Tensor sig_result = ttnn::sigmoid(input, false, output_mem_config);
+    int vector_mode = 4;
```
Comment (nvelickovicTT) [inline ttnn/cpp/ttnn/operations/eltwise/unary_backward/unary_backward.cpp:513]: Similarly here. Why not assign VectorMode::RC to male it obvious what it is, instead of a magic number? Or at least put a comment for the reader to understand what it means.

File: tt_metal/hw/ckernels/grayskull/metal/llk_api/llk_math_unary_sfpu_api.h [hunk 8]
```diff
@@ -13,6 +13,7 @@
 #include "llk_math_eltwise_unary_sfpu_rsqrt.h"
 #include "llk_math_eltwise_unary_sfpu_tiled_prod.h"
 #include "llk_math_eltwise_unary_sfpu_sigmoid.h"
+#include "llk_math_eltwise_unary_sfpu_sigmoid_appx.h"
```
[the remark this answers is not in the document]
Comment (umadevimcw) [reply tt_metal/hw/ckernels/grayskull/metal/llk_api/llk_math_unary_sfpu_api.h:16]: Removed

File: tt_metal/include/compute_kernel_api.h [hunk 9]
```diff
@@ -84,9 +84,12 @@ ALWI void rsqrt_tile(uint32_t idst) {
 /**
  * Please refer to documentation for any_init.
  */
+// clang-format on
 template <bool fast_and_approx = false>
 ALWI void sigmoid_tile_init() {
-    MATH((llk_math_eltwise_unary_sfpu_sigmoid_init<fast_and_approx>()));
+    MATH(
```
[the remark this answers is not in the document]
Comment (umadevimcw) [reply tt_metal/include/compute_kernel_api.h:90]: Updated the code

File: tt_metal/include/compute_kernel_api.h [hunk 10]
```diff
@@ -103,9 +106,11 @@ ALWI void sigmoid_tile_init() {
  * | idst            | The index of the tile in DST register buffer to perform the computation on | uint32_t | Must be less than the size of the DST register buffer | True     |
  */
 // clang-format on
-template <bool fast_and_approx = false>
+template <int vec_mode = VectorMode::RC, bool fast_and_approx = false>
 ALWI void sigmoid_tile(uint32_t idst) {
-    MATH((llk_math_eltwise_unary_sfpu_sigmoid<fast_and_approx>(idst)));
+    MATH(
```
[the remark this answers is not in the document]
Comment (umadevimcw) [reply tt_metal/include/compute_kernel_api.h:111]: Updated the code

File: ttnn/cpp/ttnn/operations/eltwise/unary/device/unary_composite_op.cpp [hunk 11]
```diff
@@ -611,7 +611,7 @@ Tensor _glu(const Tensor& input_a, int32_t dim, const std::optional<MemoryConfig
         dim = 3;
     }
     std::vector<Tensor> ab = split_tensor_for_glu(input_a, dim, output_mem_config);
-    Tensor sigmoid_b = ttnn::sigmoid(ab[1], false, output_mem_config);
+    Tensor sigmoid_b = ttnn::sigmoid(ab[1], 4, false, output_mem_config);
```
[the remark this answers is not in the document]
Comment (umadevimcw) [reply ttnn/cpp/ttnn/operations/eltwise/unary/device/unary_composite_op.cpp:614]: updated

**Review by nvelickovicTT — APPROVED** [review]

**Review by dchenTT — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/eltwise/unary/common/unary_op_utils.cpp [hunk 12]
```diff
@@ -170,6 +170,16 @@ std::pair<std::string, std::string> get_op_init_and_func_parameterized(
                 fmt::format("exp_tile_init<{}u>();", (uint32_t)param0),
                 fmt::format("exp_tile<{1}u>({0});", idst, (uint32_t)param0)};
             break;
+        case UnaryOpType::SIGMOID: {
+            uint32_t param1 = (uint32_t)params[1];
+            TT_FATAL(
```
[the remark this answers is not in the document]
Comment (umadevimcw) [reply ttnn/cpp/ttnn/operations/eltwise/unary/common/unary_op_utils.cpp:175]: Updated

File: tests/ttnn/unit_tests/operations/eltwise/test_sigmoid_vector_modes.py [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+
+
+def run_unary_test_sharded(device, hw, out_channels, vector_mode, approx_mode, ttnn_function, pcc=0.9999):
+    torch.manual_seed(0)
+
+    torch_input_tensor = torch.rand((hw, out_channels), dtype=torch.bfloat16)
+    golden_function = ttnn.get_golden_function(ttnn_function)
+    torch_output_tensor = golden_function(torch_input_tensor, device=device)
+
+    num_cores = 64
```
[the remark this answers is not in the document]
Comment (umadevimcw) [reply tests/ttnn/unit_tests/operations/eltwise/test_sigmoid_vector_modes.py]: Updated the test file
